syntax = "proto3";

service TodoService {
    rpc CreateTodo (CreateTodoRequest) returns (TodoResponse);
    rpc GetTodo (GetTodoRequest) returns (TodoResponse);
    rpc DeleteTodo (DeleteTodoRequest) returns (DeleteTodoResponse);
    rpc ListTodos (Empty) returns (TodosResponse);
}

message CreateTodoRequest {
    string title = 1;
    string message = 2;
    InnerLog inner_log = 3;
}

message GetTodoRequest {
    string id = 1;
}

message DeleteTodoRequest {
    string id = 1;
}

message TodoResponse {
    string id = 1;
    string title = 2;
    string message = 3;
    InnerLog inner_log = 4;
}

message DeleteTodoResponse {
    bool success = 1;
}

message TodosResponse {
    repeated TodoResponse todos = 1;
}

message InnerLog {
    string message = 1;
    string exception = 2;
    string stacktrace = 3;
    LogLevel level = 4;
}

enum LogLevel {
    DEBUG = 0;
    INFO = 1;
    WARNING = 2;
    ERROR = 3;
    FATAL = 4;
    CRITICAL = 5;
}

message Empty {}
